server/core_storage.go

```diff
@@ -311,9 +310,15 @@ WHERE `
 		if len(key.UserId) != 0 {
 			if uid, err := uuid.FromBytes(key.UserId); err != nil {
 				return BAD_INPUT, errors.New("Invalid user ID")
+			} else if caller != uuid.Nil && caller != uid {
+				// If the caller is a client, only allow them to write their own data.
+				return BAD_INPUT, errors.New("Clients can only remove their own data")
```

Comment:
I think we should improve the error message to "A client can only remove user-owned records.".